README.md
# legislatoR: Interface to the Comparative <img src="images/sticker.jpg" width="160" align="right" /> <br /> Legislators Database

[![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/legislatoR)](https://cran.r-project.org/package=legislatoR)
[![GitHub release version](https://img.shields.io/github/release/saschagobel/legislatoR.svg?style=flat)](https://github.com/saschagobel/legislatoR/releases)
[![CRAN_Download_Badge](https://cranlogs.r-pkg.org/badges/grand-total/legislatoR)](https://cran.rstudio.com/web/packages/legislatoR/index.html)

legislatoR is a package for the software environment R that facilitates access to the Comparative Legislators Database (CLD). The CLD includes political, sociodemographic, career, online presence, public attention, and visual information for over 67,000 contemporary and historical politicians from 16 countries. Data are also available for download in .csv and .sqlite formats at the [CLD's Dataverse](https://dataverse.harvard.edu/dataverse/cld).

## Content and data structure
The CLD covers the following countries and time periods:

| Country                              | Legislative sessions        | Politicians (unique*) | Integrated with    |
| ------------------------------------ | --------------------------- | -------------------- | ------------------ |
| Austria (Nationalrat)                | all 27<br /> (1920-2019)        | 1,923                | [ParlSpeech V2](https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/L4OAKN) (Rauh/Schwalbach 2020)      |
| Brazil (Câmara dos Deputados)        | 38-57<br /> (1947-2022)     | 3,474                |                    |
| Canada (House of Commons)            | all 44<br /> (1867-2021)        | 4,567                |                    |
| Czech Republic (Poslanecka Snemovna) | all 9<br /> (1992-2021)         | 1,124                | [ParlSpeech V1](https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/E4RSP9) (Rauh et al. 2017)          |
| France (Assemblée)                   | all 16<br /> (1958-2022)        | 4,263                |                    |
| Germany (Bundestag)                  | all 20<br /> (1949-2021)        | 4,371                | [BTVote data](https://dataverse.harvard.edu/dataverse/btvote) (Bergmann et al. 2018),<br /> [ParlSpeech V1](https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/E4RSP9) (Rauh et al. 2017),<br /> [Reelection Prospects data](https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/EBEDPI) (Stoffel/Sieberer 2017)   |
| Ireland (Dail)                       | all 33<br /> (1918-2020)        | 1,408                |	[Database of Parliamentary Speeches in Ireland](https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/6MZN76) (Herzog/Mikhaylov 2017)	|
| Israel (Knesset)                     | all 25<br /> (1949-2022)        | 1,022                |                 |
| Italy (Camera dei deputati and Senato della Repubblica) | all 19<br /> (1948-2022)  | 5,149   |                 |
| Japan (Shūgiin)                      | all 49<br /> (1890-2021)          | 6,581                |               | 
| Netherlands (Tweede Kamer)           | all 65<br /> (1815-2021)          | 1,887                |               |
| Scotland (Parliament)                | all 6<br /> (1999-2021)           | 348                  | [ParlScot](https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/EQ9WBE) (Braby/Fraser 2021)      			 |
| Spain (Congreso de los Diputados)    | all 14<br /> (1979-2019)          | 2,616           | [ParlSpeech V2](https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/L4OAKN) (Rauh/Schwalbach 2020)      |        
| Turkey (Büyük Millet Meclisi)        | all 27<br /> (1920-2018)          | 5,298                |              |
| United Kingdom (House of Commons)    | all 58<br /> (1801-2019)          | 11,321               | [EggersSpirling data](https://github.com/ArthurSpirling/EggersSpirlingDatabase) (starting from <br /> 38th session, Eggers/Spirling 2014),<br /> [ParlSpeech V1](https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/E4RSP9) (Rauh et al. 2017) | 
| United States (House and Senate)     | all 117<br /> (1789-2021)         | 12,593               | [Voteview data](https://voteview.com/data) (Lewis et al. 2019), <br /> [Congressional Bills Project data](http://www.congressionalbills.org/) (Adler/Wilkserson 2018) |
| **16**                                | **529**                     | **67,945**           | **12** 		       |

\* We only count legislators with a unique Wikipedia page or Wikidata ID. Sometimes legislators do not have either. Such cases are indicated by the string "miss" in the wikidataid or pageid.

For each legislature, the CLD holds nine tables:

1.  *Core* (sociodemographic data)
2.  *Political* (political data)
3.  *History* (full revision records of individual Wikipedia biographies)
4.  *Traffic* (daily user traffic on individual Wikipedia biographies starting from July 2007)
5.  *Social* (social media handles and personal website URLs)
6.  *Portraits* (URLs to portraits)
7.  *Offices* (public offices)
8.  *Professions* (professions)
9.  *IDs* (identifiers linking politicians to other files, databases, or websites)
 
The tables contain the following variables (see respective R help files for further details):

- *Core*: Country, Wikipedia page ID, Wikidata ID, Wikipedia Title, full name, sex, ethnicity, religion, date of birth and death, place of birth and death.
- *Political*: Wikipedia page ID, legislative session, party affiliation, lower constituency, upper constituency, constituency ID, start and end date of legislative session, period of service, majority status, leader positions.
- *History*: Wikipedia page ID, Wikipedia revision and previous revision ID, editor name/IP and ID, revision date and time, revision size, revision comment.
- *Traffic*: Wikipedia page ID, date, user traffic.
- *Social*: Wikidata ID, Twitter handle, Facebook handle, Youtube ID, Google Plus ID, Instagram handle, LinkedIn ID, personal website URL.
- *Portraits*: Wikipedia page ID, Wikipedia portrait URL.
- *Offices*: Wikidata ID, a range of offices such as attorney general, chief justice, mayor, party chair, secretary of state, etc.
- *Professions*: Wikidata ID, a range of professions such as accountant, farmer, historian, judge, mechanic, police officer, salesperson, teacher, etc.
- *IDs*: Wikidata ID, IDs for integration with various political science datsets as well as a range of other IDs such as parliamentary website IDs, Library of Congress or German National Library IDs, Notable Names Database or Project Vote Smart IDs, etc.

Note that for some legislatures or legislative periods, tables may only hold information for a subset of politicians or variables. 

The CLD comes as a relational database. This means that all tables can be joined with the *Core* table via one of two keys - the Wikipedia page ID or the Wikidata ID. These keys uniquely identify individual politicians. The figure below illustrates this structure and the CLD's content.

<p align="center">
  <img width="500" src="images/data-structure.png">
</p>

## Installation
legislatoR is available through CRAN and GitHub. To install the package from CRAN, type:

```r
install.packages("legislatoR")
```

To install the package from Github, type:

```r
devtools::install_github("saschagobel/legislatoR")
```

## Usage

A working Internet connection is required to access the CLD in R. This is because the data are stored online and not installed together with the package. The package provides table-specific function calls. These functions are named after the respective table (see [above](#content-and-data-structure)) and preceded by `get_`. To fetch  the *Core* table, use the `get_core()` function, for the *Political* table, use the `get_political()` function. Call the package help file via `?legislatoR()` to get an overview of all function calls. Tables are legislature-specific, so a three-letter country code must be passed as an argument to the function. Here is a breakdown of all country codes. You can also call the `cld_content()` function to get an overview of the CLD's scope and valid country codes.

| Country                   | Code        | Country          | Code       | Country                | Code                     |
| ------------------------- |:-----------:| ---------------- |:----------:| ---------------------- |:------------------------:|
| Austria                   | `aut`       | Ireland          | `irl`      | Spain                  | `esp`                    |
| Brazil                    | `bra`       | Israel           | `isr`      | Turkey                 | `tur`                    |
| Canada                    | `can`       | Italy            | `ita_house`/`ita_senate` | United Kingdom | `gbr`              |
| Czech Republic            | `cze`       | Japan            | `jpn`      | United States          | `usa_house`/`usa_senate` |                       | 
| France                    | `fra`       | Netherlands      | `nld`      |                        |                          |
| Germany                   | `deu`       | Scotland         | `sco`      |                        |                          |

Tables can be joined and subsetted while being fetched and memory is only allocated by the parts of a table assigned into the environment. Basic fetching, joining, and subsetting of data are illustrated below. See the Vignette [Introducing legislatoR](http://htmlpreview.github.com/?https://github.com/saschagobel/legislatoR/blob/master/vignettes/legislatoR.html) for a detailed tutorial.

```r
# load and attach legislatoR and dplyr
library(legislatoR)
library(dplyr)

# assign entire Core table for the German Bundestag into the environment
deu_politicians <- get_core(legislature = "deu")

# assign data for the 8th legislative session into the environment
deu_politicians_subset <- semi_join(x = get_core(legislature = "deu"),
				                            y = filter(get_political(legislature = "deu"), session == 8), 
			                              by = "pageid")

# join deu_politicians_subset with respective traffic on Wikipedia biographies
deu_traffic <- left_join(x = deu_politicians_subset, 
               		       y = get_traffic(legislature = "deu"), 
		                     by = "pageid")

# assign birthdate for members of the political party 'SPD' into the environment
deu_birthdates_SPD <- semi_join(x = select(get_core(legislature = "deu"), pageid, birth),
                                y = filter(get_political(legislature = "deu"), party == "SPD"),
                                by = "pageid")$birth
```

## News
See [here](NEWS.md) for details on updates.

## Glossary
See [here](GLOSSARY.md) for the full form of abbreviated country codes and party names and English translations of non-English party names.

## Sources
The CLD was predominantly built using automated data extraction techniques. See the [source code](source) and [this list](SOURCES.md) of Web sources for more details.

## Citation
Thank you for using the CLD and legislatoR! Please consider citing:

Göbel, Sascha and Simon Munzert. 2022. "[The Comparative Legislators Database](https://www.cambridge.org/core/journals/british-journal-of-political-science/article/abs/comparative-legislators-database/D28BB58A8B2C08C8593DB741F42C18B2)". *British Journal of Political Science*, 52(3), 1398-1408.

## Support
The work on this package was in part funded by the Daimler and Benz Foundation (Funding period 2017/18; project "Citizen and Elite Activity on the Wikipedia Market Place of Political Information").

Many thanks to Anna Wunderling for designing legislatoR's logo.

## Author information

**Sascha Göbel** (corresponding author and repository maintainer) <br />
Goethe University Frankfurt <br />
Faculty of Social Sciences <br />
Theodor-W.-Adorno-Platz 6 <br />
60323 Frankfurt am Main, Germany <br />
Email: sascha.goebel@soz.uni-frankfurt.de

**Simon Munzert** <br />
Hertie School


NEWS.md
# legislatoR 1.1

* Data for Canada, Czech Republic, France, Germany, Scotland, and USA updated
* Data for Brazil, Israel, Italy, Japan, Netherlands, and Turkey newly added
* All data moved to Harvard Dataverse


# legislatoR 1.0
  
* All legislative periods of Spain's Congreso de los Diputados added to the database (14 in total).
* Ongoing session of the Austrian Nationalrat added to the database (27th) and concluded session updated (26th).
* Ongoing session of the Canadian House of Commons added to the database (43th) and concluded session updated (42th).
* Ongoing session of the Irish Dáil added to the database (33th) and concluded session updated (32th).
* Ongoing session of the UK House of Commons added to the database (58th) and concluded session updated (57th).
* Irish Dáil data integrated with Database of Parliamentary Speeches in Ireland.
* Czech Poslanecka Snemovna data integrated with ParlSpeech V1.
* UK House of Commons data integrated with ParlSpeech V1.
* Spanish Congreso de los Diputados data integrated with ParlSpeech V2.
* Austrian Nationalrat data integrated with ParlSpeech V2.
* Daily Wikipedia page traffic now goes back to December 2007 (except for the US House, where file storage restrictions currently limit traffic to range back to July 2009 only).
* New `cld_content()` function that offers a quick overview of the CLD's scope and valid three-letter country codes, and helps to conventiently loop/map over legislatures and sessions.
* "Introducing legislatoR" Vignette added.
* Now available via CRAN
* New Logo/Hexsticker.


# legislatoR 0.2.0

* 1st to 37th legislative periods of UK's House of Commons added to the database (37 in total).
* Missing legislators and services variable added to all sessions of the French Assemblée.
* German Bundestag data integrated with BTVote and ParlSpeech data.
* UK House of Commons data integrated with Eggers and Spirling british political development database (starting from 38th session).
* United States House and Senate data integrated with Voteview and congressional bills project data.
* Missing Twitter handles added for the Austrian Nationalrat, Czech Poslanecka Snemovna, French Assemblée, German Bundestag, Irish Dail, UK House of Commons, and United States House and Senate (841 in total).
* Missing religious affilition added for United States House and Senate and Canada's House of Commons.
* Missing ethnicity added for Canada and UK's House of Commons.

# legislatoR 0.1.0

* Changed some function calls: `get_facial` is now `get_portrait` and `get_occupation` is now `get_profession`.
* Changed valid legislature arguments for `get_` functions. Full legislature names are no longer accepted as legislature codes. Instead, three-letter country codes such as `aut`, `can`, `cze`, `fra`, `deu`, `irl`, `sco`, `gbr`, `usa_house`, `usa_senate` must be used.
* All legislative periods of Canada's House of Commons added to the database (42 in total).
* All legislative periods of Czech Republic's Poslanecka Snemovna added to the database (8 in total).
* All legislative periods of Scotland's Parliament added to the database (5 in total).
* 38th - 57th legislative periods of UK's House of Commons added to the database (20 in total).
* Current session of the Austrian Nationalrat added to the database (26th).
* Current session of the German Bundestag added to the database (19th).
* Current session of the French Assemblée added to the database (15th).
* Current session of the United States House of representatives added to the database (116th).
* Current session of the United States Senate added to the database (116th).
* ISO 3166-1 alpha-3 three-letter country codes added to all core datasets.
* Start- and end date of sessions added to all political datasets.
* Missing legislators and services variable added to all sessions of the Irish Dáil.
* Error in United States Senate Wikipedia revision records corrected.
* Missing religious affiliation added from 1st to 18th German Bundestag using official data.
* Erroneous information on United States Senate and House members' period of service corrected.
* Additional variables on government/opposition status and leader positions added to all sessions of the Canadian House of Commons and United States Congress.
* Correction and alignment of party names across legislative periods for all parliaments.
* Emotion estimates removed from the portraits dataset.


R/cld_content.R
#' List content of the CLD
#'
#' Returns a named list of legislatures and sessions available in the CLD. This provides a quick overview of the CLD's scope and valid three-letter country codes, and helps to conventiently loop/map over legislatures and sessions.
#'
#' @param legislature An optional character string specifying one or more legislatures. Currently one of \sQuote{aut}, \sQuote{can}, \sQuote{cze}, \sQuote{esp}, \sQuote{fra}, \sQuote{deu}, \sQuote{irl}, \sQuote{sco}, \sQuote{gbr}, \sQuote{usa_house}, or \sQuote{usa_senate}. If NULL (the default), a list with all legislatures and sessions available in the CLD is returned.
#' @return A list with names being three-letter country codes and with each element containing a vector that shows the sessions available for a legislature.
#' @examples
#' # Get a list of three-letter country codes and available sessions for all countries
#' overview <- cld_content()
#' tibble::glimpse(overview)
#'
#' # Get a list of available sessions for the French Assemblée and the Irish Dail
#' sessions <- cld_content(legislature = c("fra", "irl"))
#' tibble::glimpse(sessions)
#' @export
cld_content <- function(legislature = NULL) {
  output <-  list(aut = 1:27, bra = 38:57, can = 1:44, cze = 1:9,
                  deu = 1:20, esp = 1:14, fra = 1:16, gbr = 1:58,
                  irl = 1:33, isr = 1:25, ita_house = 1:19, ita_senate = 1:19,
                  jpn = 1:49, nld = 1:65, sco = 1:6, tur = 1:27,
                  usa_house = 1:117, usa_senate = 1:117)
  if (is.null(legislature)) {
    return(output)
  } else {
    if (any(!(legislature %in% c("aut", "bra", "can", "cze",
                                 "deu", "esp", "fra", "gbr",
                                 "irl", "isr", "ita_house", "ita_senate",
                                 "jpn", "nld", "sco", "tur",
                                 "usa_house", "usa_senate")))) {
      legislature <- legislature[which(!(legislature %in% c("aut", "bra", "can", "cze",
                                                            "deu", "esp", "fra", "gbr",
                                                            "irl", "isr", "ita_house", "ita_senate",
                                                            "jpn", "nld", "sco", "tur",
                                                            "usa_house", "usa_senate")))]
      stop (paste0("\n\nPlease provide valid three-letter country codes. legislatoR does not recognize the country code or does not contain data for ",
                   paste0(
                     paste0("\"", legislature, "\""),
                     collapse = ", "),
                   ". Use `legislatoR::cld_content()` to see country codes of available legislatures."))
    }
    output <- output[legislature]
  }
  return(output)
}
